### ISSUE TITLE
Property to indicate which optional structure features are used

### ISSUE THREAD

**Issue by rartino** [Materials-Consortia/OPTiMaDe#51]
Across discussions on various aspects of the structure format, i.e., #23 resolved by PR #49 it has been discussed that by introducing optional features in separate properties, we open for the risk of naive clients that miss the use of some optional features of the format. 

In the format we ended up accepting in PR #49, the most clear example is the `assemblies` property. However, arguably also the use of disorder in structures can give similar issues. In #50 we have started discuss a separate optional property to indicate what is 'unknown' in the structural information (e.g., unknown coordinates), etc. One could make the argument here that a client is already able to filter for no optional properties by, e.g., querying for `assemblies IS NOT KNOWN` (a suggested syntax in #17). But if OPTIMaDe in later versions adds further optional properties, e.g., `magnetic_moments`, code that rely on this will break, because it will miss to filter out structures that uses this feature. 

Hence, if we continue along this direction it would be beneficial to have a property that specifies precisely what optional structure features are present for a structure. This would make it easier for clients to only query for and operate on structures they are prepared to handle (e.g., no assemblies, no unknown coordinates.)

Some options include a format like this:
```
structure_features: {
  'simple': False
  'disordered': True,
  'assembly': False,
  'unknown_positions': True
} 
```
Or this:
```
structure_features: ["disordered", "unknown_positions"]
```
Personally, I prefer the latter. Together with the syntax in #16 it will be easy to formulate sensible queries to only operate on structures with the features that one wants. E.g.,
```
filter=LENGTH structure_features=0 
```
Means to only return structures using no optional features.
```
filter=structure_features HAS ONLY "disordered" 
```
for simple structures and structures with disorder, but no other optional features.

**Comment by giovannipizzi** [comment]
I think it's a very good idea.
Another comment, though (is this already in the specs?):
all keys NOT starting with an underscore should be understood (while the custom ones with underscore, if not understood, can be ignored), and the client should "stop" or complain if it finds something it does not understand. This could be already a good method to have a safe client (and maybe this client behaviour should be specified in the specs?)

Anyway, if you think this is not enough, I like the approach `structure_features: ["disordered", "unknown_positions"]` (better than the dict, not only because of queries, but also because a client can easily have a list of "understood" features, and tell if there are new that it cannot understand.  Instead with the dict there could be a new feature but set to False and the client might not be 100% sure if "False" means that it's safe to ignore it).

**Comment by rartino** [comment]
> all keys NOT starting with an underscore should be understood (while the custom ones with underscore, if not understood, can be ignored), and the client should "stop" or complain if it finds something it does not understand. This could be already a good method to have a safe client (and maybe this client behaviour should be specified in the specs?)

If we do this, wouldn't that prevent clients from having any level of forwards-compability? Every new OPTIMaDe version will add new properties, and there is no way to tell safe properties, e.g., `symmetry_information` from unsafe properties, e.g., `unknown_positions`, so it seems clients would never be able to operate against endpoints that serve newer versions of OPTIMaDe? Or would we expect servers to know how to down-convert over large ranges of spec versions to accommodate different client versions? This seems problematic.

This discussion enters into the rather general problem of 'feature negotiation' between the client and server. But perhaps a good general solution to that problem for us are 'features'-type properties as suggested above. An `X_features` field is for listing features that, if ignored, would mean the interpretation of the data pertaining to `X` would be fundamentally incorrect. So, for structures, we would put in the spec that the client MUST consider the `structure_features` property and MUST NOT process structures with entries it does not recognize. We would then make the same setup for any other property where we foresee a similar need of covering safe forward-compability.

If we need to extend the interpretation of a property in a non-backwards-compatible way where we didn't prepare a 'features'-type field, we would, according to semantic versioning, have to increase the MAJOR version. Clients should not try to be forwards-compatible across MAJOR versions.

**Comment by giovannipizzi** [comment]
Fixed in #93

**Comment by rartino** [comment]
Indeed fixed in #93, so, lets close it.
